<a href="https://colab.research.google.com/github/2403a52210-svg/2403A52210/blob/main/NLP_presentaion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

data = pd.read_json("Sarcasm_Headlines_Dataset.json", lines=True)

print(data.head())

                                        article_link  \
0  https://www.huffingtonpost.com/entry/versace-b...   
1  https://www.huffingtonpost.com/entry/roseanne-...   
2  https://local.theonion.com/mom-starting-to-fea...   
3  https://politics.theonion.com/boehner-just-wan...   
4  https://www.huffingtonpost.com/entry/jk-rowlin...   

                                            headline  is_sarcastic  
0  former versace store clerk sues over secret 'b...             0  
1  the 'roseanne' revival catches up to our thorn...             0  
2  mom starting to fear son's web series closest ...             1  
3  boehner just wants wife to listen, not come up...             1  
4  j.k. rowling wishes snape happy birthday in th...             0  


# Task
Detect sarcasm in headlines by preparing the provided data for training, training a machine learning model, evaluating its performance, and then testing it with custom sentences.

## Prepare Data for Training

### Subtask:
Preprocess the 'headline' column by tokenizing, removing stop words, and converting the text data into numerical features suitable for machine learning, using TF-IDF.


**Reasoning**:
I will start by importing the necessary NLTK libraries, downloading the required 'punkt' and 'stopwords' corpora, and then define a preprocessing function to tokenize, lowercase, and remove stop words from the 'headline' column. Finally, I will apply this function to create a new 'processed_headline' column in the DataFrame.



In [ ]:
import pandas as pd
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from sklearn.feature_extraction.text import TfidfVectorizer

# Load the data to ensure 'data' DataFrame is defined
data = pd.read_json("Sarcasm_Headlines_Dataset.json", lines=True)

# Explicitly download necessary NLTK data to ensure full installation
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('punkt_tab') # Add this line to download the missing resource

stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    # Tokenize the text
    tokens = word_tokenize(text)
    # Convert to lowercase and remove stop words
    processed_tokens = [word.lower() for word in tokens if word.isalpha() and word.lower() not in stop_words]
    # Join the processed words back into a string
    return ' '.join(processed_tokens)

# Apply the preprocessing function to the 'headline' column
data['processed_headline'] = data['headline'].apply(preprocess_text)

print("Original headline example:", data['headline'].iloc[0])
print("Processed headline example:", data['processed_headline'].iloc[0])

# Initialize TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=5000) # Limiting features for demonstration

# Fit and transform the processed headlines into TF-IDF features
X = tfidf_vectorizer.fit_transform(data['processed_headline'])

print("\nShape of TF-IDF features (X):")
print(X.shape)
print("First 5 TF-IDF features for the first processed headline:")
print(X[0, :5].toarray())

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


Original headline example: former versace store clerk sues over secret 'black code' for minority shoppers
Processed headline example: former versace store clerk sues secret code minority shoppers

Shape of TF-IDF features (X):
(26709, 5000)
First 5 TF-IDF features for the first processed headline:
[[0. 0. 0. 0. 0.]]


# Task
Train and evaluate a sarcasm detection model using the `Sarcasm_Headlines_Dataset.json` file. The task involves loading and preprocessing the data, splitting it into training and testing sets, training a Logistic Regression model, evaluating its performance, and finally saving the trained model and TF-IDF vectorizer.

## Load and Preprocess Data

### Subtask:
Load the 'Sarcasm_Headlines_Dataset.json' file into a pandas DataFrame. Preprocess the 'headline' column by tokenizing, removing stop words, and converting the text data into numerical features using TF-IDF. This step addresses the previous file not found error and prepares the data for modeling.


**Reasoning**:
The previous code execution failed because the `Sarcasm_Headlines_Dataset.json` file was not found. Before attempting to execute the code again, it's crucial to confirm the file's presence. Since the kernel state indicates the file is still missing, I will add a new text cell to remind the user to upload the required data file as per the instructions provided in the subtask description.



**Reasoning**:
The previous attempt to add an instruction cell failed with a `SyntaxError` because the `cell_type` was incorrectly set to `python` instead of `markdown`. The content provided is indeed markdown, so I will correct the `cell_type` to `markdown` to properly display the instructions to the user.



**Reasoning**:
The previous attempt to add an instruction cell failed with a `SyntaxError` because the `cell_type` was incorrectly set to `python` instead of `markdown`. I will correct the `cell_type` to `markdown` and provide the content as a markdown block to ensure it is displayed correctly.



**Reasoning**:
The previous steps successfully loaded, preprocessed, and vectorized the data. Now, to prepare for model training, I will split the data into training and testing sets using `train_test_split` from `sklearn.model_selection`. This will create `X_train`, `X_test`, `y_train`, and `y_test` for the features and the target variable (`is_sarcastic`).



In [ ]:
from sklearn.model_selection import train_test_split

# Define target variable
y = data['is_sarcastic']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (21367, 5000)
Shape of X_test: (5342, 5000)
Shape of y_train: (21367,)
Shape of y_test: (5342,)


## Train Logistic Regression Model

### Subtask:
Initialize and train a Logistic Regression model using the preprocessed TF-IDF training data.


**Reasoning**:
I will initialize and train a Logistic Regression model using the previously generated X_train and y_train data.



In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(class_weight='balanced')
model.fit(X_train, y_train)

Logistic Regression model trained successfully.


## Evaluate Model Performance

### Subtask:
Evaluate the trained Logistic Regression model's performance on the test data using metrics such as accuracy, precision, recall, and F1-score.


**Reasoning**:
I will import the necessary evaluation metrics from `sklearn.metrics`, make predictions on the test set (`X_test`), calculate the accuracy, precision, recall, and F1-score, and then print these metrics to evaluate the model's performance as per the subtask instructions.



In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
y_pred = model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

# Print the evaluation metrics
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")

Accuracy: 0.7888
Precision: 0.7965
Recall: 0.6974
F1-score: 0.7436


## Save Model and Vectorizer

### Subtask:
Save the trained Logistic Regression model and the fitted TF-IDF vectorizer to disk as 'model.pkl' and 'vectorizer.pkl' respectively, using the pickle library.


**Reasoning**:
I will import the `pickle` module and then save the trained Logistic Regression model and the fitted TF-IDF vectorizer to disk using `pickle.dump()` as specified in the instructions.



In [ ]:
import pickle

# Save the trained Logistic Regression model
with open('model.pkl', 'wb') as f:
    pickle.dump(model, f)

# Save the fitted TF-IDF vectorizer
with open('vectorizer.pkl', 'wb') as f:
    pickle.dump(tfidf_vectorizer, f)

print("Logistic Regression model and TF-IDF vectorizer saved successfully.")

Logistic Regression model and TF-IDF vectorizer saved successfully.


## Final Task

### Subtask:
Summarize the completion of the model training, evaluation, and saving process, confirming that the model and vectorizer are ready for future use.


## Summary:

### Q&A
The task was to train and evaluate a sarcasm detection model, then save the model and vectorizer. This goal was successfully achieved.

### Data Analysis Key Findings
*   **Data Preparation:** The `Sarcasm_Headlines_Dataset.json` file was successfully loaded, and its 'headline' column underwent thorough preprocessing, including tokenization, stop-word removal, and lowercasing. The cleaned text was then converted into numerical TF-IDF features, resulting in a feature matrix `X` of shape (26709, 5000). The data was split into training and testing sets, with `X_train` having a shape of (21367, 5000) and `X_test` having a shape of (5342, 5000).
*   **Model Training:** A Logistic Regression model was initialized with `max_iter=1000` and successfully trained using the preprocessed TF-IDF training data (`X_train` and `y_train`).
*   **Model Performance:** The trained Logistic Regression model achieved an accuracy of 0.7888, precision of 0.7965, recall of 0.6974, and an F1-score of 0.7436 on the test set.
*   **Model Persistence:** Both the trained Logistic Regression model and the fitted TF-IDF vectorizer were successfully serialized and saved to disk as 'model.pkl' and 'vectorizer.pkl', respectively, ensuring they are ready for future use.

### Insights or Next Steps
*   The model demonstrates reasonable performance in sarcasm detection, with an F1-score of approximately 0.74, indicating a balanced performance between precision and recall.
*   Consider exploring more advanced NLP techniques or models (e.g., neural networks like LSTMs or Transformers) to potentially improve the recall score, as it is slightly lower than precision, which might be important for identifying sarcastic content.


# Task
*   **Correct cell type**: Change the cell type of `a5c3020e` from `code_cell` to `text_cell` to properly display the markdown instructions.

## Correct cell type

### Subtask:
Change the cell type of cell `a5c3020e` from code to text.


## Summary:

### Data Analysis Key Findings
*   The cell with ID `a5c3020e` was successfully changed from a code cell to a text cell.

### Insights or Next Steps
*   This correction was necessary to properly display markdown instructions within the notebook.


# Task
**Correct cell types**: Change the cell type of cells `a5c3020e`, `7921dd78`, and `46da5343` from `code_cell` to `text_cell`.

## Correct Cell Types

### Subtask:
Change the cell type of cells `a5c3020e`, `7921dd78`, and `46da5343` from `code_cell` to `text_cell` to resolve the `SyntaxError` and properly display the instructions.
